### ISSUE TITLE
No redirect to info.json service

### ISSUE THREAD

**Issue by beaudet** [ruven/iipsrv#126]
The IIIF API requires that for non-level 0 compliance, requests for the image itself should redirect to the info.json service.  How would you like to handle this Ruven?

**Comment by ahankinson** [comment]
If it's of any use, I have the following two nginx rules for handling this part of the spec:

```
 location ~ ^/iiif/image/(?<iiif>.*)(?<ext>jpx|jp2|tif)$ {
        return 303 /iiif/image/$iiif$ext/info.json;
    }

rewrite ^/iiif/image/(?<iiif>.*)(?<action>default.jpg|info.json)$ /iip/iipsrv.fcgi?IIIF=%2f$iiif$action last;
```

**Comment by beaudet** [comment]
yeah, I'm trying to avoid having to handle this at the web server level - I see there is support for this redirect in IIIF.cc but it's not working properly when the image ID has slashes in it, e.g. 

https://media.nga.gov/iiif/public/objects/1/0/6/3/8/2/106382-primary-0-nativeres.ptif

It's getting confused parsing the URL.  I'm working on a patch to first stat check the file portion and if it exists redirect to info.json.

**Comment by ahankinson** [comment]
The nice thing about having it at the web server level is the ability to bring in things like regex matches.  Will you be able to 'opt-out' of this behaviour?

**Comment by beaudet** [comment]
Not without also opting-out of the IIIF image API specs.  This only applies to the IIIF handler by the way.

**Comment by ahankinson** [comment]
I was just curious if this would conflict with any IIIF rewrite rules already in place to deal with this.

**Comment by beaudet** [comment]
I doubt it as the web server gets first crack at handling URLs  so any existing rules are likely to take precedence and still work.

**Comment by regisrob** [comment]
I notice that this issue has not been solved (I'm using IIP 1.1 version). The `baseUriRedirect` to the info.json is not handled properly when there is a slash in the identifier slot, e.g. "https://example.org/iiif/folder/image.jp2" does not redirect to "https://example.org/iiif/folder/image.jp2/info.json". And it is a requirement at level 1 of the Image API https://iiif.io/api/image/2.1/compliance/#http-features

I'm seeing this in the logs:
```
Full Request is IIIF=folder/image.jp2
[1/1]: Command / Argument is IIIF : folder/image.jp2
IIIF handler reached
IIIF :: URL decoded to folder/image.jp2
IIIF: Not enough parameters
Sending HTTP 400 Bad Request
```
I guess IIP should handle this by itself (and not at web server level), given that it has already been partially implemented and does work correctly for image identifiers without slash.

What do you think @ruven?

**Comment by ruven** [comment]
iipsrv 1.1 does in fact redirect requests for the image itself to the info.json URL, but as you correctly point out, not for paths containing slashes.

However, the IIIF specification says the identifier must be url-encoded (https://iiif.io/api/image/2.1/#uri-encoding-and-decoding), so your example, which is an unencoded URL, should return an error.

Unfortunately, even if you correctly URL encode the path, iipsrv still doesn't handle this situation correctly, so I'll try to add a fix to make sure this works as expected.

**Comment by regisrob** [comment]
thanks @ruven, you're right about the unencoded URL with respect to the IIIF spec, even though some servers do interpret non-encoded slashes in the identifier slot (thus not strictly conforming to the spec). This is the case for Loris, and Gallica as well (e.g. https://gallica.bnf.fr/iiif/ark:/12148/btv1b10038246f/f1/info.json).

I've noticed that IIP has the same behavior when the identifier is passed in the URL as a parameter, like `https://example.org/iipsrv/iipsrv.fcgi?iiif=folder/image.jp2/info.json`: in this case there is no need to percent-encode the slash (contrary to a URL like `https://example.org/iipsrv/folder/image.jp2/info.json`), but maybe that's precisely because the identifier slot is passed as a parameter and not in the URL path...

**Comment by ruven** [comment]
> even though some servers do interpret non-encoded slashes in the identifier slot (thus not strictly conforming to the spec)

iipsrv tries to follow the spec, but tries to be as forgiving as possible with user input! Things still work even if the client isn't 100% conformant.

**Comment by ahankinson** [comment]
IIP should behave consistently with url-encoded and non-url-encoded slashes in the identifier slot, since clients do not always handle this correctly (or worse, transparently "fix" it for you). 

AFAIK, IIIF URLs are well-defined, so it would be possible to detect the identifier from the remainder of parsing the path: 
 - Suffix can either be `/x,y,w,h/w,h/rot/quality.format` or `/info.json`
 - Prefix is the hostname and an optional path prefix*

Everything else is the identifier. If you get something that does not have a known suffix, and the prefix matches the configured value, then you know that you have an identifier and can send a redirect to the `/info.json` response.

* I think IIP lacks the ability to set a path prefix (everything between the hostname and the identifier) but if it had that it would save a lot of mucking about with trying to determine whether the URL-encoded slashes (or non-URL-encoded slashes) are part of the identifier or not. 

** I'm still in favour of suggesting that complex path handling happens upstream at the HTTP server level. They do this sort of thing very easily and very well. IIP is really great at being an image server, but putting in URL handling and rewriting and redirects in a sufficiently robust manner to meet all use cases seems like a bit much.
